recaf-core/src/main/java/me/coley/recaf/parse/JavaParserHelper.java

```diff
@@ -167,45 +171,42 @@ private boolean isInvalidCompilationUnit(CompilationUnit unit) {
-				ItemInfo value = JavaParserResolving.of(symbolSolver, node);
-				if (value == null) {
-					Optional<Node> parent = node.getParentNode();
-					if (parent.isPresent()) {
-						// Check if the parent item can be resolved
-						node = parent.get();
-						continue;
-					} else {
-						// No parent, end the loop
-						break;
-					}
-				} else if (node instanceof SimpleName && value instanceof FieldInfo) {
-					// In some obfuscated cases, a method name may be selected as an AST 'SimpleName'.
-					// These typically resolve to fields of matching names.
+		if (unit == null) return Optional.empty();
+		Node node = getNodeAtLocation(line, column, unit);
+		if (node instanceof Expression) {
+			Optional<Number> nbOpt = evaluate((Expression) node);
+			if (nbOpt.isPresent()) {
+				return Optional.of(new ParseHitResult(new LiteralExpressionInfo(nbOpt.get(), (Expression) node), node));
+			}
+		}
+		while (node != null && JavaParserResolving.isNodeResolvable(node)) {
+			ItemInfo value = JavaParserResolving.of(symbolSolver, node);
+			if (value == null) {
+				Optional<Node> parent = node.getParentNode();
+				// Check if the parent item can be resolved
+				if (parent.isEmpty()) break; // No parent, end the loop
+				node = parent.get();
+				continue;
+			} else if (node instanceof SimpleName && value instanceof FieldInfo) {
+				// In some obfuscated cases, a method name may be selected as an AST 'SimpleName'.
+				// These typically resolve to fields of matching names.
+				Optional<Node> parent = node.getParentNode();
+				if (parent.isPresent()) {
+					ItemInfo info = JavaParserResolving.of(symbolSolver, parent.get());
 					// If the parent of the node is a method of the same name, we probably meant to
-					// yield the method info, not a field by the same name.
-					FieldInfo fieldInfo = (FieldInfo) value;
-					if (node.hasParentNode()) {
-						ItemInfo info = JavaParserResolving.of(symbolSolver, node.getParentNode().get());
-						if (info instanceof MethodInfo) {
-							MethodInfo methodInfo = (MethodInfo) info;
-							if (methodInfo.getName().equals(fieldInfo.getName())) {
-								value = info;
-							}
-						}
+					if (info instanceof MethodInfo && info.getName().equals(value.getName())) {
+						// yield the method info, not a field by the same name.
+						value = info;
 					}
 				}
-				return Optional.of(new ParseHitResult(value, node));
-			}
-			// Handle edge cases like package import names.
-			ItemInfo value = JavaParserResolving.ofEdgeCases(typeSolver, node);
-			if (value != null) {
-				return Optional.of(new ParseHitResult(value, node));
 			}
+			return Optional.of(new ParseHitResult(value, node));
 		}
-		return Optional.empty();
+		// Handle edge cases like package import names.
+		ItemInfo value = JavaParserResolving.ofEdgeCases(typeSolver, node);
+		if (value == null) return Optional.empty();
+
+		return Optional.of(new ParseHitResult(value, node));
```

Comment:
Mostly a note to self: Lots of changes here, will need to validate there aren't any odd regressions in context resolution.

Reply:
I refactored the method and made more "early return" to keep it more flat.